### DISCUSSION TITLE
[DataGrid] NRE when deleting item right after editing

### DISCUSSION THREAD

**Discussion by jp2masa** [AvaloniaUI/Avalonia#9303] [Q&A] [answered]
**Describe the bug**

Editing an item and triggering it's removal from the data grid items list results in an NRE when committing the edit.

**To Reproduce**

1. Start from the Avalonia MVVM template, and install data grid
2. On `MainWindow.xaml`:

    ```xaml
    <DataGrid AutoGenerateColumns="True" Items="{Binding Items}" />
    ```
3. On `MainWindowViewModel.cs`:

    ```cs
    public class MainWindowViewModel : ViewModelBase
    {
        public sealed class Item : ViewModelBase
        {
            private string _value;

            public Item(MainWindowViewModel owner, string value)
            {
                _value = value;

                this.WhenAnyValue(x => x.Value).Subscribe(_ => owner.Items.Remove(this));
            }

            public string Value
            {
                get => _value;
                set => this.RaiseAndSetIfChanged(ref _value, value);
            }
        }

        public MainWindowViewModel()
        {
            Items.Add(new Item(this, "A"));
            Items.Add(new Item(this, "B"));
        }

        public ObservableCollection<Item> Items { get; } = new ObservableCollection<Item>();
    }
    ```
4. Run the app, and edit one of the values, then commit the edit, and it throws an NRE.

**Expected behavior**

The item should be successfully removed from the list.

**Desktop (please complete the following information):**

- OS: Windows 11
- Version: 11.0.0-preview2

**Comment by timunie** [comment, marked as answer]
I think it works if you use the `Dispatcher` to remove the items. That way you can wait for Avalonia to commit the changes before removing the item. I've used the following item view model for my tests: 

```cs
public sealed class Item : ViewModelBase
{
    private string _value;

    // Added two private fields which we need to access later. 
    private MainWindowViewModel owner;
    private bool _init;
    
    public Item(MainWindowViewModel owner, string value)
    {
        this.owner = owner;
        _value = value;
        
        this.WhenAnyValue(x => x.Value)
            .Subscribe(_ => RemoveMe());

        // RxUI is calling WhenAnyValue in the constructor for some reason. We don't want this, so we wait for the model to be ready
        _init = true;
    }

    
    // I made a function of it. Feel free to move it to your lambda if you want to.
    private void RemoveMe()
    {
        if (!_init) return;
        Dispatcher.UIThread.Post(() => owner.Items.Remove(this));
    }
    
    public string Value
    {
        get => _value;
        set => this.RaiseAndSetIfChanged(ref _value, value);
    }
}
```

tbh I don't consider this a bug. Therefore I move this into a discussion. If you are okay with the solution showed above, please mark it as the answer for others.

**Reply by jp2masa** [reply]
This may not be a really useful case, but an NRE is always a bug, it should throw some other exception.

**Reply by timunie** [reply]
You're probably right. If you have some time to setup your sample in our sandbox app (clone the source code) you can set a break point where the NRE is thrown. There you can probably add a null check and throw another exception. If you are happy with it, open a PR to fix it 😁

**Reply by timunie** [reply]
Hey @jp2masa the NRE occurs in these lines: https://github.com/AvaloniaUI/Avalonia/blob/af1c316f53e246cdbe4fbd4983d677b667b45d66/src/Avalonia.Controls.DataGrid/DataGrid.cs#L4124-L4143

Which kind of Exception do you find useful in this case?
